### PR TITLE
Implement WheelEvent interface

### LINKED ISSUE #22843: Implement WheelEvent interface
Spec: https://w3c.github.io/uievents/#events-wheelevents
Docs: https://doc.servo.org/script/dom/index.html#adding-a-new-dom-interface

I see `[2019-02-06T22:41:04Z ERROR script::dom::bindings::error] Error at https://assets-prod.reticulum.io/assets/js/hub-01a8e131ea6755144653.js:1:376343 WheelEvent is not defined` when creating a Hubs room.

### PR DESCRIPTION
**Pull request by robert-snakard**
Created a new dom interface: "WheelEvent" and added WPT tests to confirm the interface works. To do this I had to do the following:

- Create a new `WheelEvent` dom interface. It can be found in `script/dom/wheelevent.rs` and `dom/webidls/WheelEvent.webidl`
- Add a new `WheelEvent` option to the compositor's `CompositorEvent` enum
- Add a new `Wheel` option to the compositor's `WindowEvent` enum
- Add a new `WheelDelta` type to the `script_traits` module
- Modify the `scroll_event` logic. Now we send a `WheelEvent` before scrolling. Repeat: we send the WheelEvent notification BEFORE we send the scroll delta.
- Add two manual wpt tests to the `uievents/order-of-events` test collection

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #22843 (GitHub issue number if applicable)

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23154)

### CODE REVIEW

**Comment by robert-snakard** [comment]
@jdm

**Comment by jdm** [comment]
@robert-snakard Looks like this needs to be rebased in a way that doesn't include commits which are already present on master.

**Comment by robert-snakard** [comment]
This should just be master with 2 commits on top...

**Comment by jdm** [comment]
It's possible that because the @bors-servo merge commits are missing it is not actually equivalent to master. I recommend doing a hard reset to master and cherry-picking your two commits, then force pushing.

**Comment by robert-snakard** [comment]
@jdm Updated the branch version and tested on my machine, the patch still works. Do I need to do anything else?

**Comment by jdm** [comment]
No, we just need to review these changes now.

**Comment by jdm** [comment]
@paulrouget Review ping?

**Comment by robert-snakard** [comment]
This failed because tests that were supposed to fail are now passing, looking for the idlharness now to update it

**Comment by jdm** [comment]
This failure also occurred in the other test job:
```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Constructed WheelEvent timestamp should be high resolution and have the same time origin as performance.now()", 
    "test": "/dom/events/Event-timestamp-high-resolution.html", 
    "line": 80347, 
    "action": "test_result", 
    "expected": "FAIL"
}
```

**Comment by jdm** [comment]
@paulrouget Reminder review ping!

**Comment by paulrouget** [comment]
Sorry, don't know how I missed that. Looking today.

File: components/script_traits/lib.rs [hunk 1]
```diff
@@ -460,6 +460,19 @@ pub enum MouseEventType {
...
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
+pub struct WheelDelta {
+    /// Delta in the left/right direction
+    pub x: f64,
+    /// Delta in the up/down direction
+    pub y: f64,
+    /// Delta in the direction going into/out of the screen
+    pub z: f64,
+    /// Mode to measure the floats in
+    pub mode: u32,
```
Comment (paulrouget) [inline components/script_traits/lib.rs]: Maybe use an enum + FromPrimitive.

File: ports/servo/glutin_app/window.rs [hunk 2]
```diff
@@ -495,14 +495,21 @@ impl Window {
...
             } => {
-                let (mut dx, mut dy) = match delta {
-                    MouseScrollDelta::LineDelta(dx, dy) => (dx, dy * LINE_HEIGHT),
+                let (mut dx, mut dy, mode) = match delta {
+                    MouseScrollDelta::LineDelta(dx, dy) => (dx as f64, (dy * LINE_HEIGHT) as f64, 0x01),
                     MouseScrollDelta::PixelDelta(position) => {
                         let position =
                             position.to_physical(self.device_hidpi_factor().get() as f64);
-                        (position.x as f32, position.y as f32)
+                        (position.x as f64, position.y as f64, 0x00)
```
Comment (paulrouget) [inline ports/servo/glutin_app/window.rs]: 0x00 and 0x01 look like magic numbers. Can you use an enum?

File: components/script/dom/document.rs [hunk 3]
```diff
@@ -1265,6 +1268,56 @@ impl Document {
...
+        debug!("{}: on {:?}", wheel_event_type_string, node.debug_str());
+
+        // https://w3c.github.io/uievents/#event-wheelevents
+        let event = WheelEvent::new(
+            &self.window,
+            DOMString::from(wheel_event_type_string),
+            EventBubbles::Bubbles,
+            EventCancelable::Cancelable,
+            Some(&self.window),
+            1,
```
Comment (paulrouget) [inline components/script/dom/document.rs]: Why 1? Isn't it supposed to be 0?

File: components/script/dom/document.rs [hunk 4]
```diff
@@ -1265,6 +1268,56 @@ impl Document {
...
+        );
+
+        let event = event.upcast::<Event>();
+        event.set_trusted(true);
+
+        let target = node.upcast();
+        event.fire(target);
+
+        self.window
+            .reflow(ReflowGoal::Full, ReflowReason::WheelEvent);
```
Comment (paulrouget) [inline components/script/dom/document.rs]: Why is a reflow necessary here?

File: components/script/dom/document.rs [hunk 5]
```diff
@@ -1265,6 +1268,56 @@ impl Document {
...
 
+    #[allow(unsafe_code)]
+    pub fn handle_wheel_event(
+        &self,
+        js_runtime: *mut JSRuntime,
+        delta: WheelDelta,
+        client_point: Point2D<f32>,
+        node_Address: Option<UntrustedNodeAddress>,
+    ) {
+        let wheel_event_type_string = "wheel".to_owned();
```
Comment (paulrouget) [inline components/script/dom/document.rs]: Is it necessary to have a String here? Doesn't `DOMString::new` work with a str?

**Review by paulrouget — COMMENTED** [review]
Looks good to me, modulo some nits.

I'm not very familiar with the DOM/Webidl code. @jdm can you take a look at components/script/dom/wheelevent.rs - it looks good to me, but I'd like you to take a look as well.

File: components/script_traits/lib.rs (hunk 1 above)
Comment (robert-snakard) [reply components/script_traits/lib.rs]: Not sure exactly what you mean. How would I implement FromPrimitive for this vector? Also are you thinking something like this?
```
enum WheelDelta {
    ModeDeltaPixel { x: f64, y: f64, z: f64 },
    ModeDeltaLine { x: f64, y: f64, z: f64 },
    ModeDeltaPage { x: f64, y: f64, z: f64 },
}
```

File: components/script/dom/document.rs (hunk 4 above)
Comment (robert-snakard) [reply components/script/dom/document.rs]: It's not, at one point I was trying to move scrolling to inside the wheel event

File: ports/servo/glutin_app/window.rs (hunk 2 above)
Comment (robert-snakard) [reply ports/servo/glutin_app/window.rs]: Yea, Will Fix

File: components/script/dom/document.rs (hunk 5 above)
Comment (robert-snakard) [reply components/script/dom/document.rs]: nope, `DOMString::new()` only creates an empty domstring. I could do `DOMString::new(); DOMString::push_str(&mut self, &str)`. I used `from_string` cause that's what the mouse event used.

File: components/script/dom/document.rs (hunk 3 above)
Comment (robert-snakard) [reply components/script/dom/document.rs]: yes. Will Fix

**Comment by jdm** [comment]
DOM code looks good to me!

File: components/script_traits/lib.rs (hunk 1 above)
Comment (paulrouget) [reply components/script_traits/lib.rs]: I meant for the `mode` field.

File: ports/servo/glutin_app/window.rs (hunk 2 above)
Comment (robert-snakard) [reply ports/servo/glutin_app/window.rs]: @jdm the webidl file generates an enum for us in the `components/script` crate. I've tried importing to `window.rs` it with `use script::dom::bindings` and `use components::script::dom:: bindings` but both say "use of undeclared type or module". Is it possible to access this crate from the `ports/servo/glutin_app` directory?

Comment (jdm) [reply ports/servo/glutin_app/window.rs]: You will want to create an independent enum in components/script_traits and convert between it and the generated one inside the script/ code.

**Comment by jdm** [comment]
@paulrouget Is there anything else needed here besides rebasing?

**Comment by jdm** [comment]
Oh, sorry, I just noticed the outstanding code changes from the last review. Carry on.

File: components/script_traits/lib.rs (hunk 1 above)
Comment (robert-snakard) [reply components/script_traits/lib.rs]: Changed to `pub mode: WheelMode` and `enum WheelMode { Pixel = 0x00 }`. Didn't do a FromPrimitive impl

**Comment by robert-snakard** [comment]
Fixed @paulrouget's issues, updated to most recent master.

File: components/script/dom/window.rs [hunk 6]
```diff
@@ -2169,6 +2170,7 @@ fn debug_reflow_events(id: PipelineId, reflow_goal: &ReflowGoal, reason: &Reflow
         ReflowReason::FirstLoad => "\tFirstLoad",
         ReflowReason::KeyEvent => "\tKeyEvent",
         ReflowReason::MouseEvent => "\tMouseEvent",
+        ReflowReason::WheelEvent => "\tWheelEvent",
```
Comment (paulrouget) [inline components/script/dom/window.rs]: Is that necessary?

**Comment by paulrouget** [comment]
Other than my comment, it looks good.

**Comment by jdm** [comment]
Yeah, I think we can collapse the MouseEvent and WheelEvent reflow reasons together.

**Comment by paulrouget** [comment]
There's no reflow on wheel event.

**Comment by paulrouget** [comment]
Looks good to me. Please squash the commits.

@bors-servo try

**Comment by paulrouget** [comment]
@bors-servo try-

See Travis output.

File: components/script/dom/document.rs [hunk 7]
```diff
@@ -1196,6 +1199,53 @@ impl Document {
             .reflow(ReflowGoal::Full, ReflowReason::MouseEvent);
     }
 
+    #[allow(unsafe_code)]
+    pub fn handle_wheel_event(
+        &self,
+        js_runtime: *mut JSRuntime,
+        delta: WheelDelta,
+        client_point: Point2D<f32>,
+        node_Address: Option<UntrustedNodeAddress>,
```
Comment (CYBAI) [inline components/script/dom/document.rs]: nit: let's use snake case which means `node_address`.

File: components/script/dom/document.rs [hunk 8]
```diff
@@ -1196,6 +1199,53 @@ impl Document {
...
+                .filter_map(DomRoot::downcast::<Element>)
+                .next()
+        });
+
+        let el = match el {
+            Some(el) => el,
+            None => return,
+        };
+
+        let node = el.upcast::<Node>();
```
Comment (CYBAI) [inline components/script/dom/document.rs]: Not sure if you'd think it' easier to read if we use `if let` to extract the element?

```rust
let node = if let Some(el) = el {
  el.upcast::<Node>();
} else {
  return;
}
```

Comment (robert-snakard) [reply components/script/dom/document.rs]: Leaving this one the way it is

**Comment by robert-snakard** [comment]
@paulrouget @jdm merged the commits. Do you want the wpt tests as a separate commit or part of this one? What about the third commit, preference for making it part of commit 1 or commit 2?

**Comment by jdm** [comment]
Leaving the tests in a separate commit is fine, and merging the test updates into the first commit would be good. You may need to reset the branch to a good master revision and then cherry-pick your commits on top of it, though.

**Comment by robert-snakard** [comment]
>> Build execution time has reached the maximum allowed time for your plan (100 minutes).

@jdm what do I do about this?

**Comment by jdm** [comment]
Nothing. That's fine.

**Comment by paulrouget** [comment]
Conflicting files: tests/wpt/metadata/uievents/idlharness.window.js.ini

**Comment by paulrouget** [comment]
@robert-snakard can you rebase your PR?

**Comment by robert-snakard** [comment]
i rebased right before pushing so there shouldn't be a problem.

I can't get in front of a computer till late tonight. I'll rebase then.

**Comment by jdm** [comment]
```

  ▶ Unexpected subtest result in /uievents/idlharness.window.html:
  └ PASS [expected FAIL] MouseEvent interface: new WheelEvent("event") must inherit property "buttons" with the proper type
```

**Comment by robert-snakard** [comment]
Missed that one when merging with the mousebuttons patch. Will fix tonight.
Is there any way to run these tests on my local machine before I push?

**Comment by jdm** [comment]
You can run individual tests or directories with `./mach test-wpt /uievents/idlharness.window.html` or `./mach test-wpt /uievents/` (for example), but running all of the tests locally is a mostly futile effort because there are so many of them.

**Comment by jdm** [comment]
Looks like these tests as well:
```
  ▶ Unexpected subtest result in /dom/events/Event-subclasses-constructors.html:
  └ PASS [expected FAIL] WheelEvent constructor (no argument)

  ▶ Unexpected subtest result in /dom/events/Event-subclasses-constructors.html:
  └ PASS [expected FAIL] WheelEvent constructor (undefined argument)

  ▶ Unexpected subtest result in /dom/events/Event-subclasses-constructors.html:
  └ PASS [expected FAIL] WheelEvent constructor (null argument)

  ▶ Unexpected subtest result in /dom/events/Event-subclasses-constructors.html:
  └ PASS [expected FAIL] WheelEvent constructor (empty argument)

  ▶ Unexpected subtest result in /dom/events/Event-subclasses-constructors.html:
  └ PASS [expected FAIL] WheelEvent constructor (argument with default values)
```
Otherwise it should be ready :)

**Comment by robert-snakard** [comment]
Thanks for finishing the merge up for me @jdm / @paulrouget

**Comment by jdm** [comment]
Thanks for doing the hard work :)

**Comment by robert-snakard** [comment]
Haha, we must have different definitions of hard work, that merge took forever!

**Pull request merged**